File: simo-player/tools/pack_run.js
#!/usr/bin/env node
/* Contributor packer: run SUMO, pack FCD+summary to BLGR, emit .simo.json.
 *
 * Port of tools/pack_run.py. Single scenario:
 *     pack_run.js --net n.xml --rou r.xml --scenario today -o out.simo.json
 *
 * A/B (each scenario has its own net + routes):
 *     pack_run.js --scenario today=t.net.xml:t.rou.xml \
 *                 --scenario proposed=p.net.xml:p.rou.xml -o out.simo.json
 */
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { findSumo as findSumoBinary, geom, geoLock } from './sumo_geom.js';
import {
  buildTypeMap, readDemand, packFcd, packStats, readFcd, readSummary,
} from './blgr_pack.js';

/* Exact flag list from video_capture.py:34-38 — locked by test/tools.test.js
 * (seed 42, step-length 1). */
export const SUMO_FLAGS = [
  '--fcd-output', '{fcd}',
  '--fcd-output.geo', 'false',
  '--summary-output', '{summ}',
  '--seed', '42',
  '--step-length', '1',
  '--time-to-teleport', '300',
  '--max-depart-delay', '1800',
  '-e', '{end}',
  '--no-step-log', '--no-warnings',
];

function fail(msg) {
  process.stderr.write(msg + '\n');
  process.exit(2);
}

export function usage() {
  return `usage: pack_run.js [-h] --scenario SCENARIO -o OUTPUT [--net NET] [--rou ROU]
                    [--cfg CFG] [--end END] [--sumo SUMO]

Pack a SUMO run into .simo.json

options:
  --net NET               SUMO network file (single-scenario form)
  --rou ROU               SUMO routes file (single-scenario form)
  --cfg CFG               SUMO config (.sumocfg) supplying net/rou
  --scenario SCENARIO     Scenario key, or key=net.xml:rou.xml for A/B.
                          Repeatable.
  -o OUTPUT, --output OUTPUT
                          Output .simo.json path
  --end END               Simulation end time (seconds)
  --sumo SUMO             Path to sumo binary (auto-discovered if omitted)`;
}

export function parseArgs(argv) {
  const opts = { scenario: [], end: 900 };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const next = () => {
      i += 1;
      if (i >= argv.length) fail(`argument ${a}: expected one argument`);
      return argv[i];
    };
    if (a === '--help' || a === '-h') {
      process.stdout.write(usage() + '\n');
      process.exit(0);
    } else if (a === '--net') opts.net = next();
    else if (a === '--rou') opts.rou = next();
    else if (a === '--cfg') opts.cfg = next();
    else if (a === '--scenario') opts.scenario.push(next());
    else if (a === '-o' || a === '--output') opts.output = next();
    else if (a === '--end') {
      const v = Number(next());
      if (!Number.isFinite(v)) fail('argument --end: invalid int value');
      opts.end = Math.trunc(v);
    } else if (a === '--sumo') opts.sumo = next();
    else fail(`unrecognized argument: ${a}`);
  }
  if (!opts.scenario.length) fail('the following arguments are required: --scenario');
  if (!opts.output) fail('the following arguments are required: -o/--output');
  return opts;
}

/* Normalize --scenario values into [(key, net_path, rou_path), ...].
 *
 * Two forms, never mixed:
 *   bare keys        -> global --net/--rou (single scenario only)
 *   key=net:rou      -> per-scenario inputs (A/B) */
export function parseScenarioSpecs(scenarioArgs, net, rou) {
  const specs = [];
  const keyed = scenarioArgs.map((s) => s.includes('='));
  if (keyed.some((k) => k) && !keyed.every((k) => k)) {
    throw new Error('mixing bare scenario keys and key=net:rou specs');
  }
  if (keyed.every((k) => k)) {
    for (const s of scenarioArgs) {
      const eq = s.indexOf('=');
      const key = s.slice(0, eq);
      const paths = s.slice(eq + 1);
      const c = paths.lastIndexOf(':');
      const n = c === -1 ? '' : paths.slice(0, c);
      const r = c === -1 ? '' : paths.slice(c + 1);
      if (!key || !n || !r) {
        throw new Error(`bad --scenario spec: ${JSON.stringify(s)} `
          + '(want key=net.xml:rou.xml)');
      }
      specs.push([key, n, r]);
    }
  } else {
    if (!net || !rou) {
      throw new Error('bare --scenario keys require --net and --rou');
    }
    if (scenarioArgs.length > 1) {
      throw new Error(
        'multiple scenarios share one net/rou — an A/B run needs '
        + 'per-scenario specs: --scenario today=t.net.xml:t.rou.xml '
        + '--scenario proposed=p.net.xml:p.rou.xml');
    }
    specs.push([scenarioArgs[0], net, rou]);
  }
  const keys = specs.map(([k]) => k);
  if (new Set(keys).size !== keys.length) {
    throw new Error(`duplicate scenario keys: ${JSON.stringify(keys)}`);
  }
  return specs;
}

/* Run SUMO with exact flags from video_capture.py:34-38. */
export function runSumo(sumo, net, rou, fcdOut, summOut, endTime, sumoHome) {
  const cmd = [sumo, '-n', net, '-r', rou];
  for (const flag of SUMO_FLAGS) {
    if (flag === '{fcd}') cmd.push(fcdOut);
    else if (flag === '{summ}') cmd.push(summOut);
    else if (flag === '{end}') cmd.push(String(endTime));
    else cmd.push(flag);
  }
  const r = spawnSync(cmd[0], cmd.slice(1), {
    encoding: 'utf8',
    env: sumoHome ? { ...process.env, SUMO_HOME: sumoHome } : process.env,
  });
  if (r.status !== 0) {
    throw new Error(`SUMO failed: ${(r.stderr || '').slice(-800)}`);
  }
  return true;
}

/* Assemble the .simo.json dict.
 *
 * scenarios_data: {key: {frames, stats, geometry, geo, demand}} where geo is
 * sumo_geom.geoLock()'s {latlngMap, utm} or null. Scenario entries follow
 * the catalog shape: lanes/arms/phases/stops/links + frames/stats (base64)
 * + latlngMap/utm/geoLocked when provenance exists. Top level adds demand
 * (max scenario rate) and peakServed (max final arrived across scenarios)
 * — matching build_mock's catalog semantics. */
export function buildPack(scenariosData, nFrames) {
  const outScenarios = {};
  /* running min/max — Math.min(...xs) overflows the call stack once the
   * packed record count passes V8's spread-argument limit (~65k; a
   * gridlocked corridor run over 900 frames reaches millions). */
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  let anchor = null;
  const demands = [];
  let peakServed = 0;
  for (const [key, d] of Object.entries(scenariosData)) {
    if (d.demand != null) demands.push(d.demand);
    if (d.stats && d.stats.length) {
      peakServed = Math.max(peakServed, d.stats[d.stats.length - 1][0]);
    }
    const geometry = d.geometry;
    const geo = d.geo || null;
    const latlngMap = geo ? geo.latlngMap : null;
    const utm = geo ? geo.utm : null;
    const framesBlob = packFcd(d.frames);
    const statsBlob = packStats(d.stats);
    const sc = { title: key.toUpperCase(),
      lanes: geometry.lanes, arms: geometry.arms || {},
      phases: geometry.phases || [],
      stops: geometry.stops || {}, links: geometry.links || {},
      frames: Buffer.from(framesBlob).toString('base64'),
      stats: Buffer.from(statsBlob).toString('base64') };
    if (latlngMap) {
      sc.latlngMap = latlngMap;
      sc.geoLocked = true;
      if (anchor === null) {
        const o = latlngMap.orig;
        anchor = [(o[0] + o[2]) / 2, (o[1] + o[3]) / 2];
      }
    }
    if (utm) {
      sc.utm = utm;
      sc.geoLocked = true;
    }
    outScenarios[key] = sc;
    for (const frame of d.frames) {
      for (const v of frame) {
        if (v[1] < minX) minX = v[1];
        if (v[2] < minY) minY = v[2];
        if (v[1] > maxX) maxX = v[1];
        if (v[2] > maxY) maxY = v[2];
      }
    }
  }
  const bounds = Number.isFinite(minX)
    ? [minX, minY, maxX, maxY] : [0, 0, 1, 1];
  return { nFrames, bounds, anchor,
    demand: demands.length ? Math.max(...demands) : null,
    peakServed,
    scenarios: outScenarios };
}

export async function main(argv) {
  const args = parseArgs(argv);

  let net = args.net, rou = args.rou;
  if (args.cfg) {
    const cfg = fs.readFileSync(args.cfg, 'utf8');
    const cfgDir = path.dirname(path.resolve(args.cfg));
    const nf = /<net-file\b[^>]*\bvalue\s*=\s*"([^"]+)"/.exec(cfg);
    const rf = /<route-files\b[^>]*\bvalue\s*=\s*"([^"]+)"/.exec(cfg);
    if (!nf || !rf) fail('error: net-file/route-files missing in config');
    net = path.join(cfgDir, nf[1]);
    rou = path.join(cfgDir, rf[1]);
  }

  let specs;
  try {
    specs = parseScenarioSpecs(args.scenario, net, rou);
  } catch (e) {
    fail(`error: ${e.message}`);
  }

  const sumo = args.sumo || findSumoBinary();
  if (!sumo) fail('error: sumo not found. Set --sumo or $SUMO_HOME');

  let sumoHome = null;
  if (process.env.SUMO_HOME) {
    sumoHome = process.env.SUMO_HOME;
  } else if (sumo.startsWith('/Library/Frameworks/EclipseSUMO.framework/')) {
    sumoHome = '/Library/Frameworks/EclipseSUMO.framework/'
      + 'Versions/Current/EclipseSUMO/share/sumo';
  }

  const scenariosData = {};
  const td = fs.mkdtempSync(path.join(os.tmpdir(), 'simo-pack-'));
  for (const [key, netPath, rouPath] of specs) {
    const fcdOut = path.join(td, `${key}-fcd.xml`);
    const summOut = path.join(td, `${key}-sum.xml`);
    process.stdout.write(`  Running SUMO for '${key}' (${path.basename(netPath)})...\n`);
    runSumo(sumo, netPath, rouPath, fcdOut, summOut, args.end, sumoHome);
    const tmap = buildTypeMap(rouPath);
    const [frames, nVeh] = await readFcd(key, td, args.end, tmap);
    const stats = await readSummary(key, td, args.end);
    process.stdout.write(`    ${nVeh} vehicles, ${frames.length} frames, `
      + `${Object.keys(tmap).length} vTypes\n`);
    scenariosData[key] = {
      frames, stats,
      geometry: geom(netPath),
      geo: geoLock(netPath),
      demand: readDemand(rouPath),
    };
  }

  const out = buildPack(scenariosData, args.end);
  fs.writeFileSync(args.output, JSON.stringify(out));
  const kb = (fs.statSync(args.output).size / 1024).toFixed(1);
  process.stdout.write(`Wrote ${args.output} (${kb} KB)`
    + (out.anchor ? ' [geo-locked]' : '') + '\n');
}

if (process.argv[1] && path.resolve(process.argv[1]) === path.resolve(new URL(import.meta.url).pathname)) {
  main(process.argv.slice(2)).catch((e) => fail(String((e && e.stack) || e)));
}

File: simo-player/tools/pack_run.py
#!/usr/bin/env python3
"""Contributor packer: run SUMO, pack FCD+summary to BLGR, emit .simo.json.

Single scenario:
    pack_run.py --net n.xml --rou r.xml --scenario today -o out.simo.json

A/B (each scenario has its own net + routes):
    pack_run.py --scenario today=t.net.xml:t.rou.xml \
                --scenario proposed=p.net.xml:p.rou.xml -o out.simo.json
"""
import argparse
import base64
import json
import os
import struct
import subprocess
import sys
import tempfile

# Add tools to path for imports
TOOLS_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, TOOLS_DIR)

import sumo_geom
import blgr_pack


SUMO_FLAGS = [
    "--fcd-output", "{fcd}",
    "--fcd-output.geo", "false",
    "--summary-output", "{summ}",
    "--seed", "42",
    "--step-length", "1",
    "--time-to-teleport", "300",
    "--max-depart-delay", "1800",
    "-e", "{end}",
    "--no-step-log", "--no-warnings"
]


def parse_args():
    p = argparse.ArgumentParser(description="Pack SUMO run into .simo.json")
    p.add_argument("--net", help="SUMO network file (single-scenario form)")
    p.add_argument("--rou", help="SUMO routes file (single-scenario form)")
    p.add_argument("--cfg", help="SUMO config (.sumocfg) supplying net/rou")
    p.add_argument("--scenario", required=True, action="append",
                   help="Scenario key, or key=net.xml:rou.xml for A/B. "
                        "Repeatable.")
    p.add_argument("-o", "--output", required=True, help="Output .simo.json path")
    p.add_argument("--end", type=int, default=900,
                   help="Simulation end time (seconds)")
    p.add_argument("--sumo", help="Path to sumo binary (auto-discovered if omitted)")
    return p.parse_args()


def parse_scenario_specs(scenario_args, net, rou):
    """Normalize --scenario values into [(key, net_path, rou_path), ...].

    Two forms, never mixed:
      bare keys        -> global --net/--rou (single scenario only)
      key=net:rou      -> per-scenario inputs (A/B)
    """
    specs = []
    keyed = ["=" in s for s in scenario_args]
    if any(keyed) and not all(keyed):
        raise ValueError("mixing bare scenario keys and key=net:rou specs")
    if all(keyed):
        for s in scenario_args:
            key, _, paths = s.partition("=")
            n, _, r = paths.rpartition(":")
            if not key or not n or not r:
                raise ValueError(f"bad --scenario spec: {s!r} "
                                 "(want key=net.xml:rou.xml)")
            specs.append((key, n, r))
    else:
        if not net or not rou:
            raise ValueError("bare --scenario keys require --net and --rou")
        if len(scenario_args) > 1:
            raise ValueError(
                "multiple scenarios share one net/rou — an A/B run needs "
                "per-scenario specs: --scenario today=t.net.xml:t.rou.xml "
                "--scenario proposed=p.net.xml:p.rou.xml")
        specs.append((scenario_args[0], net, rou))
    keys = [k for k, _n, _r in specs]
    if len(set(keys)) != len(keys):
        raise ValueError(f"duplicate scenario keys: {keys}")
    return specs


def find_sumo(sumo_arg):
    if sumo_arg:
        return sumo_arg
    return sumo_geom.find_sumo()


def run_sumo(sumo, net, rou, fcd_out, summ_out, end_time, sumo_home):
    """Run SUMO with exact flags from video_capture.py:34-38."""
    cmd = [sumo, "-n", net, "-r", rou]
    for flag in SUMO_FLAGS:
        if flag == "{fcd}":
            cmd.append(fcd_out)
        elif flag == "{summ}":
            cmd.append(summ_out)
        elif flag == "{end}":
            cmd.append(str(end_time))
        else:
            cmd.append(flag)
    env = dict(os.environ, SUMO_HOME=sumo_home) if sumo_home else os.environ
    result = subprocess.run(cmd, capture_output=True, text=True, env=env)
    if result.returncode != 0:
        raise RuntimeError(f"SUMO failed: {result.stderr[-800:]}")
    return True


def read_fcd(fcd_path, n_frames, type_map=None):
    """Parse FCD XML into per-frame vehicle lists."""
    import xml.etree.ElementTree as ET
    tmap = type_map or {}
    per = [[] for _ in range(n_frames)]
    ids = {}
    for _ev, el in ET.iterparse(fcd_path, events=("end",)):
        if el.tag == "timestep":
            t = int(float(el.get("time")))
            if 0 <= t < n_frames:
                per[t] = [(ids.setdefault(v.get("id"), len(ids)),
                           round(float(v.get("x")) * 10),
                           round(float(v.get("y")) * 10),
                           int(float(v.get("angle")) / 2) % 180,
                           min(255, int(float(v.get("speed")) * 8)),
                           tmap.get(v.get("type"), 0)) for v in el]
            el.clear()
    return per, len(ids)


def read_summary(summ_path, n_frames):
    """Parse summary XML into per-frame stat rows."""
    import xml.etree.ElementTree as ET
    per = [(0, 0, 0, 0, 0)] * n_frames
    out = list(per)
    for _ev, el in ET.iterparse(summ_path, events=("end",)):
        if el.tag == "step":
            t = int(float(el.get("time")))
            if 0 <= t < n_frames:
                run, halt = int(el.get("running")), int(el.get("halting"))
                out[t] = (min(65535, int(el.get("arrived"))),
                          min(65535, max(0, run - halt)), min(65535, halt),
                          min(65535, int(el.get("waiting"))),
                          min(65535, int(el.get("teleports"))))
            el.clear()
    return out


def build_pack(scenarios_data, n_frames):
    """Assemble the .simo.json dict.

    scenarios_data: {key: {"frames": per-frame vehicle lists,
                           "stats": per-frame [5] rows,
                           "geometry": sumo_geom.geom() dict,
                           "geo": (latlngMap, utm) or None,
                           "demand": veh/hr from read_demand, or None}}
    Scenario entries follow the catalog shape: lanes/arms/phases/stops/links
    + frames/stats (base64) + latlngMap/utm/geoLocked when provenance exists.
    Top level adds demand (max scenario rate) and peakServed (max final
    arrived across scenarios) — matching build_mock's catalog semantics.
    """
    out_scenarios = {}
    xs, ys, anchor = [], [], None
    demands, peak_served = [], 0
    for key, d in scenarios_data.items():
        if d.get("demand") is not None:
            demands.append(d["demand"])
        if d["stats"]:
            peak_served = max(peak_served, d["stats"][-1][0])
        geom = d["geometry"]
        geo = d.get("geo")
        latlng_map, utm = geo if geo else (None, None)
        frames_blob = blgr_pack.pack_fcd(d["frames"])
        stats_blob = blgr_pack.pack_stats(d["stats"])
        sc = {"title": key.upper(),
              "lanes": geom["lanes"], "arms": geom.get("arms", {}),
              "phases": geom.get("phases", []),
              "stops": geom.get("stops", {}), "links": geom.get("links", {}),
              "frames": base64.b64encode(frames_blob).decode("ascii"),
              "stats": base64.b64encode(stats_blob).decode("ascii")}
        if latlng_map:
            sc["latlngMap"] = latlng_map
            sc["geoLocked"] = True
            if anchor is None:
                o = latlng_map["orig"]
                anchor = [(o[0] + o[2]) / 2, (o[1] + o[3]) / 2]
        if utm:
            sc["utm"] = utm
            sc["geoLocked"] = True
        out_scenarios[key] = sc
        for frame in d["frames"]:
            for v in frame:
                xs.append(v[1])
                ys.append(v[2])
    bounds = [min(xs), min(ys), max(xs), max(ys)] if xs else [0, 0, 1, 1]
    return {"nFrames": n_frames, "bounds": bounds, "anchor": anchor,
            "demand": max(demands) if demands else None,
            "peakServed": peak_served,
            "scenarios": out_scenarios}


def main():
    args = parse_args()

    net, rou = args.net, args.rou
    if args.cfg:
        import xml.etree.ElementTree as ET
        root = ET.parse(args.cfg).getroot()
        cfg_dir = os.path.dirname(os.path.abspath(args.cfg))
        nf = root.find(".//net-file")
        rf = root.find(".//route-files")
        if nf is None or rf is None:
            sys.exit("error: net-file/route-files missing in config")
        net = os.path.join(cfg_dir, nf.get("value"))
        rou = os.path.join(cfg_dir, rf.get("value"))

    try:
        specs = parse_scenario_specs(args.scenario, net, rou)
    except ValueError as e:
        sys.exit(f"error: {e}")

    sumo = find_sumo(args.sumo)
    if not sumo:
        sys.exit("error: sumo not found. Set --sumo or $SUMO_HOME")

    sumo_home = None
    if os.environ.get("SUMO_HOME"):
        sumo_home = os.environ["SUMO_HOME"]
    elif sumo.startswith("/Library/Frameworks/EclipseSUMO.framework/"):
        sumo_home = ("/Library/Frameworks/EclipseSUMO.framework/"
                     "Versions/Current/EclipseSUMO/share/sumo")

    scenarios_data = {}
    with tempfile.TemporaryDirectory() as td:
        for key, net_path, rou_path in specs:
            fcd_out = os.path.join(td, f"{key}-fcd.xml")
            summ_out = os.path.join(td, f"{key}-sum.xml")
            print(f"  Running SUMO for '{key}' ({os.path.basename(net_path)})...")
            run_sumo(sumo, net_path, rou_path, fcd_out, summ_out,
                     args.end, sumo_home)
            tmap = blgr_pack.build_type_map(rou_path)
            frames, n_veh = read_fcd(fcd_out, args.end, tmap)
            stats = read_summary(summ_out, args.end)
            print(f"    {n_veh} vehicles, {len(frames)} frames, "
                  f"{len(tmap)} vTypes")
            scenarios_data[key] = {
                "frames": frames, "stats": stats,
                "geometry": sumo_geom.geom(net_path),
                "geo": sumo_geom.geo_lock(net_path),
                "demand": blgr_pack.read_demand(rou_path),
            }

    out = build_pack(scenarios_data, args.end)
    with open(args.output, "w") as f:
        json.dump(out, f, separators=(",", ":"))
    print(f"Wrote {args.output} ({os.path.getsize(args.output)/1024:.1f} KB)"
          + (" [geo-locked]" if out["anchor"] else ""))


if __name__ == "__main__":
    main()